### ISSUE TITLE
[1.3-modules-beta.4] modules not working with coffeescript

### ISSUE THREAD

**Issue by mjmasn** [meteor/meteor#6000]
The compiled coffeescript isn't being run through the modules parser as .js code would be. When using backticks in a .coffee file to be able to use import/export, the actual import/export statements are left in the final compiled .js file instead of being converted to require(x) calls.

@benjamn let me know if you need a repro repo of this, but it seems unnecessary here given that it's an issue for any coffeescript code that uses backtick-ed import/export statements.

The issue occurs at app level and with packages.

**imports/somelib.coffee:**
```coffeescript
hello = 'world'
`export { hello as default }`
```

**test.coffee:**
```coffeescript
`import hello from './imports/somelib'`
console.log hello
```

**package.js:**
```javascript
Package.describe({
  name: 'testpkg',
  version: '0.0.1',
  summary: '',
  git: '',
  documentation: 'README.md'
});

Package.onUse(function(api) {
  api.versionsFrom('1.3-modules-beta.4');
  api.use('ecmascript');
  api.use('coffeescript');
  api.use('modules');
  api.addFiles('test.coffee');
});
```
**error:**
```
=> Exited with code: 8
          
/Users/mike/Code/coffeetest/.meteor/local/build/programs/server/packages/testpkg.js:28
import hello from './imports/somelib';                               // 1
^^^^^^
SyntaxError: Unexpected reserved word
    at /Users/mike/Code/coffeetest/.meteor/local/build/programs/server/boot.js:241:30
    at Array.forEach (native)
    at Function._.each._.forEach (/Users/mike/.meteor/packages/meteor-tool/.1.1.11-modules.5.fk4njb++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/server-lib/node_modules/underscore/underscore.js:79:11)
    at /Users/mike/Code/coffeetest/.meteor/local/build/programs/server/boot.js:137:5
```

**Comment by mquandalle** [comment]
The `import`/`export` syntax is specific to ES6 transpilers and it won’t work if you just put it on the generated JavaScript file. AFAIK Coffeescript doesn’t specify a built-in syntax for that, however you should be able to use the `require` function that is used under the hood by the transpiled JavaScript code:

```coffeescript
hello = require('./imports/somelib')
```

And for the export:

```coffeescript
exports.hello = 'world'
```

**Comment by mjmasn** [comment]
I'm aware of that - what I'm asking for is for the resulting JavaScript to also be transpiled with the new modules processor if there is an import or export statement in the output. I personally don't much mind having to use require and exports as it's what I'm used to anyway, but then we start getting a split between how to do Meteor with Javascript and Meteor with CoffeeScript (and admittedly this is already the case with '@'-ing global variables and the package-scoped share object).

In theory Meteor's CoffeeScript processor could detect plain import/export statements and automatically backtick them first to make life easier, but I can see this causing problems at a later stage if ES6 support *was* eventually added to CoffeeScript (an unlikely scenario as far as I can tell).

Maybe I should just learn to love curly-braces and semi-colons until someone builds a modern CoffeeScript clone...

**Comment by mquandalle** [comment]
> Maybe I should just learn to love curly-braces and semi-colons until someone builds a modern CoffeeScript clone...

Well curly-braces should be enough, as semi-colons aren’t required and will be inserted by Babel anyway ;-)

**Comment by avital** [comment]
It's not just modules that you can't access with CoffeeScript's backticks -- you can't access all of ES6.

Running all of Babel through the output of CoffeeScript would be slower, and -- does anyone do this in their non-Meteor projects?

I think it could be fine to just say, "until CoffeeScript supports modules, you'll have to use `require`"

**Comment by mjmasn** [comment]
Fair enough, I'm happy to use require. I wouldn't hold my breath for ES6 support in CoffeeScript though :(

**Comment by GeoffreyBooth** [comment]
I suggest we rename this issue “CoffeeScript code wrapped in backticks not passed to ecmascript” and treat it as a real issue. I think when both the `coffeescript` **and** the `ecmascript` packages are included in a project, `.coffee` files should first be processed by `coffeescript` and the output processed by `ecmascript`. Yes it’s slow, but it may be required if a certain Meteor (or more likely, React/JSX) feature is only achievable with ES2015. For people who don’t want the slowdown of two compilation steps, all they would need to do is remove `ecmascript` from their project—then backticked code would be passed through without further processing.

**Comment by rriemann** [comment]
Should this be closed?

**Comment by GeoffreyBooth** [comment]
I guess when 4191d01877533d7935df3843ae41fee40f35459a is released, probably as part of 1.3.3, this can be closed.
